vertical-pod-autoscaler/recommender/base/histogram.go

```diff
@@ -0,0 +1,82 @@
+package base
+
+// Histogram represents an approximate distribution of some variable.
+type Histogram interface {
+	// Returns an approximation of the given percentile of the distribution.
+	// Note: the argument passed to Percentile() is a number between 0 and 1.
+	// For example 0.5 corresponds to the median and 0.9 to the 90th percentile.
+	// If the histogram is empty, Percentile() returns 0.0.
+	Percentile(percentile float64) float64
+
+	// Add a sample with a given value and weight. A sample can have negative
+	// weight, as long as the total weight of samples with the given value is
+	// not negative.
+	AddSample(value float64, weight float64)
+
+	// Returns true if the histogram is empty.
+	Empty() bool
+}
+
+// Simple bucket-based implementation of the histogram. Resolution of the histogram
+// depends on the options. Samples are rounded down to the bottom of the bucket.
+// There's no interpolation within buckets.
+type SimpleHistogram struct {
+	options *HistogramOptions	// Bucketing scheme.
+	buckets []float64	// Weight of samples in each bucket.
+	totalWeight float64	// Weight of samples in all buckets.
```

Comment:
Each sample has a weight (e.g. for a resource usage sample it could be the length of the measurement interval in seconds, etc).
Each bucket holds the total weight of samples that fall into this bucket.
In the simplest case you could just always use weight = 1.0, in which case each bucket will hold the number of samples.